ddtrace/profiling/collector/_traceback.pyx

```diff
@@ -49,11 +93,25 @@ cpdef pyframe_to_frames(frame, max_nframes):
     :return: The serialized frames and the number of frames present in the original traceback."""
     frames = []
     nframes = 0
-    while frame is not None:
-        nframes += 1
-        if len(frames) < max_nframes:
-            code = frame.f_code
-            lineno = 0 if frame.f_lineno is None else frame.f_lineno
-            frames.append((code.co_filename, lineno, code.co_name, _extract_class_name(frame)))
-        frame = frame.f_back
+    # Python 3.11 moved PyFrameObject members to internal C API and cannot be directly accessed
+    IF PY_MAJOR_VERSION > 3 or (PY_MAJOR_VERSION == 3 and PY_MINOR_VERSION >= 11):
+        cdef PyFrameObject* pyframe = <PyFrameObject*> frame
+        while pyframe != NULL:
+            nframes += 1
+            if len(frames) < max_nframes:
+                code = PyFrame_GetCode(pyframe)
+                lineno = PyFrame_GetLineNumber(pyframe)
+                lineno = 0 if lineno is None else lineno
+                frames.append(((<object>code).co_filename, lineno, (<object>code).co_name, _extract_class_name(<object>pyframe)))
+                Py_XDECREF(<PyObject*>code)
+            pyframe = PyFrame_GetBack(pyframe)
+            # FIXME: Where to Py_XDECREF(pyframe)?
```

Comment:
you might need to do something like:

```python

cdef PyFrameObject* cur_frame = pyframe
pyframe = PyFrame_GetBack(cur_frame)
Py_XDECREF(cur_frame)
```

? and then a final `Py_XDECREF(pyframe)` after the while ?

Reply:
I think this makes sense. Trying that now.

Reply:
I would have thought that
~~~
Py_XDECREF(pyframe)
pyframe = PyFrame_GetBack(cur_frame)
~~~
is enough, but @brettlangdon's suggestion works too.